### ISSUE TITLE
Decompression vs Sequential Read

### ISSUE THREAD

**Issue by Stack-Attack** [webdataset/webdataset#217]
Hello~

**Comment by tmbdev** [comment]
Yes, I understand the problem. Trying to read multiple shards simultaneously from the same rotational drive is very slow.

If it's just a single shard, why not just copy it to a RAM file system or SSD instead?

If you're actually trying to read multiple shards on the same rotational disks in parallel, you need to make the accesses sequential.

The easiest way of dealing with this is actually to just run a web server on your hard drive and configure it to serve a maximum of one simultaneous requests. Then you can access the shards through http://localhost/... The web server is fast compared to the disk I/O.

You can also use a small script that ensures sequential access via file locking:

```
# usage: cat1 file
exec 17>>/tmp/job.lock
flock -x 17
mbuffer -m 2G -i "$1"
# exiting closes the file descriptor and releases the lock
```

Then you can access the file with "pipe:cat1 /data/shard{000000..000999}.tar.gz"

(I don't guarantee that the script works as written, but it should give you an idea. You probably want to add error handling as well.)

With this in place, you can then have as many workers as you need to handle the CPU load.

This is still not quite ideal, since you may not want to buffer every shard fully 

Finally, you can have a single WebDataset reader process but use Python `multiprocessing` to execute code over that in parallel.

```
def process1(sample):
    ...
def process(source):
    ...
    pool = mp.Pool(...)
    ...
    for sample in pool.imap_async(source, process1):
        yield sample

ds = WebDataset(...).compose(process)
```

**Comment by Stack-Attack** [comment]
@tmbdev Thanks for the thorough response! If I run into this constraint again in the future I'll likely decode/decompress samples in some parallel manner through the python method you mentioned. It's not ideal, but seems simple and elegant enough to keep the HDD saturated.

You mention 'the web server is fast compared to the disk I/O'. Without the strict locking mechanism you mention, can I ask for a brief explanation on why this is?
